src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -121,11 +124,104 @@ protected override void RemoveItem(int index)
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            T[] oldItems = new T[Items.Count];
+            if (Items is List<T> list)
+            {
+                list.CopyTo(oldItems);
+            }
+            else
+            {
+                for (int i = 0; i < oldItems.Length; i++)
+                {
+                    oldItems[i] = this[i];
+                }
+            }
+
+            bool ignore = _skipRaisingEvents;
+            if (!ignore)
+            {
+                _skipRaisingEvents = true;
+            }
+
+            try
+            {
+                base.RemoveItemsRange(index, count);
+            }
+            finally
+            {
+                if (!ignore)
+                {
+                    _skipRaisingEvents = false;
+                }
+            }
+            
+            if (count > 0 && !_skipRaisingEvents)
+            {
+                T[] removedItems = new T[count];
+                for (int i = 0; i < count; i++)
+                {
+                    removedItems[i] = oldItems[index + i];
+                }
+
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a collection of items is added to list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void ReplaceItemsRange(int index, int count, IEnumerable<T> collection)
+        {
+            CheckReentrancy();
+
+            _skipRaisingEvents = true;
+
+            IList itemsToReplace = new List<T>();
+            for (int i = index; i < count; i++)
+            {
+                itemsToReplace.Add(this[i]);
+            }            
+
+            try
+            {
+                base.ReplaceItemsRange(index, count, collection);
+            }
+            finally
+            {
+                _skipRaisingEvents = false;
+            }
+            
             OnCountPropertyChanged();
             OnIndexerPropertyChanged();
-            OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            OnCollectionChanged(NotifyCollectionChangedAction.Replace, itemsToReplace, (IList)collection, index);
```

Comment:
`(IList)collection` is not going to work if the passed-in `collection` doesn't implement `IList`.

Instead, this needs to be checked and converted, if needed, before being passed to `OnCollectionChanged`, e.g.:

```c#
IList newItems = collection is IList list ? list : new List<T>(collection);

OnCountPropertyChanged();
OnIndexerPropertyChanged();
OnCollectionChanged(NotifyCollectionChangedAction.Replace, itemsToReplace, newItems, index);
```

Reply:
Thanks for the recommendation. I added this in.